File: Leetcode/FindtheMaximumSumofNodeValues.cpp
class Solution {
   public:
    long long inf = 9999999999999999;
    void dfs(int cur,
             int k,
             const vector<vector<int>>& adj,
             const vector<int>& nums,
             vector<array<long long, 2>>& dp,
             vector<bool>& used) {
        used[cur] = true;

        auto csize = adj[cur].size();
        auto cdp_prev = array<long long, 2>{0, -inf};
        auto cdp = array<long long, 2>{0, 0};

        for (int i = 0; i < csize; ++i) {
            auto to = adj[cur][i];
            if (used[to]) {
                continue;
            } else {
                dfs(to, k, adj, nums, dp, used);

                cdp[0] = max(cdp_prev[0] + dp[to][0], cdp_prev[1] + dp[to][1]);
                cdp[1] = max(cdp_prev[1] + dp[to][0], cdp_prev[0] + dp[to][1]);

                swap(cdp, cdp_prev);
            }
        }

        dp[cur][0] =
            max(cdp_prev[1] + (nums[cur] ^ k), cdp_prev[0] + nums[cur]);
        dp[cur][1] =
            max(cdp_prev[0] + (nums[cur] ^ k), cdp_prev[1] + nums[cur]);
    }

    long long maximumValueSum(vector<int>& nums,
                              int k,
                              vector<vector<int>>& edges) {
        int n = nums.size();

        vector<vector<int>> adj(n, vector<int>());
        for (int i = 0; i < edges.size(); ++i) {
            auto u = edges[i][0];
            auto v = edges[i][1];

            adj[u].push_back(v);
            adj[v].push_back(u);
        }

        vector<array<long long, 2>> dp(n, {0, 0});
        vector<bool> used(n, false);

        dfs(0, k, adj, nums, dp, used);

        return dp[0][0];
    }
};

File: Leetcode/FindtheMaximumSumofNodeValues.rs
// https://leetcode.com/problems/find-the-maximum-sum-of-node-values

use std::mem::swap;

struct Solution {}

impl Solution {
    pub fn maximum_value_sum(nums: Vec<i32>, k: i32, edges: Vec<Vec<i32>>) -> i64 {
        let k = k as i64;

        let nums = nums.into_iter().map(|x| x as i64).collect::<Vec<_>>();

        let edges = edges.into_iter().fold(vec![vec![]; nums.len()], |mut edges, edge| {
            let v = edge[0] as usize;
            let u = edge[1] as usize;
            edges[v].push(u);
            edges[u].push(v);

            edges
        });

        let root = edges
            .iter()
            .enumerate()
            .find(|(_, edges)| edges.len() == 1)
            .unwrap()
            .0;

        #[rustfmt::skip]
        fn dfs(cur: usize, par: i32, k: i64, nums: &Vec<i64>, edges: &Vec<Vec<usize>>) -> [i64; 2] {
            let mut ldp  = [0, i64::MIN];
            let mut nldp = [0, 0];

            for &to in edges[cur].iter() {
                if to as i32 == par {
                    continue;
                }

                let res = dfs(to, cur as i32, k, nums, edges);

                nldp[0] = (
                    ldp[0] + nums[to] + res[0]).max(
                    ldp[0] + (nums[to] ^ k) + res[1]).max(
                    ldp[1] + (nums[to] ^ k) + res[0]).max(
                    ldp[1] + nums[to] + res[1]);
                
                nldp[1] = (
                    ldp[1] + nums[to] + res[0]).max(
                    ldp[1] + (nums[to] ^ k) + res[1]).max(
                    ldp[0] + (nums[to] ^ k) + res[0]).max(
                    ldp[0] + nums[to] + res[1]);
                    
                swap(&mut ldp, &mut nldp);
            }

            ldp
        }

        let res = dfs(root, -1, k, &nums, &edges);

        (res[0] + nums[root]).max(res[1] + (nums[root] ^ k))
    }
}